ote/src/clj/ote/services/service_search.clj

```diff
@@ -30,13 +30,37 @@
 (defn- ids [key query-result]
   (into #{} (map key) query-result))
 
-(defn- operation-area-ids [db operation-area]
+(defn- services-operating-in
+  "Returns ids of services which operate in the given areas"
+  [db operation-area]
   (when (seq operation-area)
     (ids
-      ::search/transport-service-id
-      (specql/fetch db ::search/operation-area-facet
-                    #{::search/transport-service-id}
-                    {::search/operation-area (op/in (map str/lower-case operation-area))}))))
+     :id
+     (service-ids-by-operation-areas db {:operation-area operation-area}))))
+
+(defn match-quality [intersection difference]
+  "Returns match quality of a search result to the operation-area it was searched against. Greater return values are worse matches.
+   Negative return value is an invalid match"
+  (if (pos? intersection) (/ difference intersection) -1))
+
+(defn sort-by-match-quality
+  "Sorts results by the match quality. Removes results with invalid match qualities"
+  [results spatial-diffs]
+  (let [quality-by-id (zipmap (map :id spatial-diffs) (map #(match-quality (:intersection %) (:difference %)) spatial-diffs))]
+    (->> results
+         (map #(assoc % :match-quality (get quality-by-id (::t-service/id %))))
```

Comment:
No need to change this now, but it seems to me that this could be simpler.
If (match-quality) fn would take row as parameter and would add quality as map parameter when quality is positive and nil if negative. Then we would not need zip map and this map could be a keep fn, with call to match-quality.

Reply:
Currently I have a collection of results and a collection of spatial differences. How can I pick the correct spatial difference for a result without the map? I think this is where some of the idea of first get the keys, then get the values is beginning to break down a little bit. I definitely could change map -> filter to one keep call, that much I can understand.